File: backend/app/services/permission_service.py
from sqlalchemy.orm import Session

from app.models.permission import Permission
from app.repositories.permission_repository import PermissionRepository
from app.schemas.permission import (
    PermissionCreate,
    PermissionUpdate,
)


class PermissionService:

    def __init__(
        self,
        db: Session,
    ):
        self.repository = PermissionRepository(db)

    def get_permissions(self):
        return self.repository.get_all()

    def get_permission_by_id(
        self,
        permission_id: int,
    ):
        return self.repository.get_by_id(
            permission_id
        )

    def create_permission(
        self,
        data: PermissionCreate,
    ):
        existing = self.repository.get_by_name(
            data.name
        )

        if existing:
            raise ValueError(
                "Permission already exists"
            )

        permission = Permission(
            name=data.name
        )

        return self.repository.create(
            permission
        )

    def update_permission(
        self,
        permission_id: int,
        data: PermissionUpdate,
    ):
        permission = self.repository.get_by_id(
            permission_id
        )

        if permission is None:
            return None

        existing = self.repository.get_by_name(
            data.name
        )

        if (
            existing is not None
            and existing.id != permission_id
        ):
            raise ValueError(
                "Permission already exists"
            )

        return self.repository.update(
            permission,
            data.name,
        )

    def delete_permission(
        self,
        permission_id: int,
    ):
        permission = self.repository.get_by_id(
            permission_id
        )

        if permission is None:
            return False

        return self.repository.delete(
            permission
        )

File: frontend/src/api/permissionService.js
import api from "./axios";

/**
 * Get All Permissions
 */
export async function getPermissions() {
  const response = await api.get("/permissions/");
  return response.data;
}

/**
 * Get Permission By ID
 */
export async function getPermission(permissionId) {
  const response = await api.get(
    `/permissions/${permissionId}`
  );

  return response.data;
}

/**
 * Create Permission
 */
export async function createPermission(
  permissionData
) {
  const response = await api.post(
    "/permissions/",
    permissionData
  );

  return response.data;
}

/**
 * Update Permission
 */
export async function updatePermission(
  permissionId,
  permissionData
) {
  const response = await api.put(
    `/permissions/${permissionId}`,
    permissionData
  );

  return response.data;
}

/**
 * Delete Permission
 */
export async function deletePermission(
  permissionId
) {
  const response = await api.delete(
    `/permissions/${permissionId}`
  );

  return response.data;
}
